diff --git a/src/GNUmakefile b/src/GNUmakefile
index b74b3a3..468c000 100644
--- a/src/GNUmakefile
+++ b/src/GNUmakefile
@@ -63,8 +63,13 @@ endef
 
 $(eval $(call define_from_arch.h,,HAS_LOADER_32BIT))
 
+ifdef X86_32BIT
+  HAS_LOADER_32BIT =
+  LOADER_FLAGS_X86_32 = -m32
+endif
 ifdef HAS_LOADER_32BIT
   OBJECTS += loader/loader-m32-wrapped.o
+  LOADER_FLAGS_X86_32 =
 endif
 
 .DEFAULT_GOAL = proot
@@ -163,8 +168,8 @@ LOADER$1_OBJECTS = loader/loader$1.o loader/assembly$1.o
 $(eval $(call define_from_arch.h,$1,LOADER_ARCH_CFLAGS))
 $(eval $(call define_from_arch.h,$1,LOADER_ADDRESS))
 
-LOADER_CFLAGS$1  += -fPIC -ffreestanding $(LOADER_ARCH_CFLAGS$1)
-LOADER_LDFLAGS$1 += -static -nostdlib -Wl$(BUILD_ID_NONE),-Ttext=$(LOADER_ADDRESS$1),-z,noexecstack
+LOADER_CFLAGS$1  += $(LOADER_FLAGS_X86_32) -fPIC -ffreestanding $(LOADER_ARCH_CFLAGS$1)
+LOADER_LDFLAGS$1 += $(LOADER_FLAGS_X86_32) -static -nostdlib -Wl$(BUILD_ID_NONE),-Ttext=$(LOADER_ADDRESS$1),-z,noexecstack
 
 loader/loader$1.o: loader/loader.c
 	@mkdir -p $$(dir $$@)
diff --git a/src/execve/exit.c b/src/execve/exit.c
index 475e002..9c4ca69 100644
--- a/src/execve/exit.c
+++ b/src/execve/exit.c
@@ -420,8 +420,10 @@ void translate_execve_exit(Tracee *tracee)
 		poke_reg(tracee, STATE_FLAGS, 0);
 
 #if defined(ARCH_ARM_EABI) && defined(__thumb__)
+#ifdef LEAVE_ARM_THUMB_MODE
 		/* Leave ARM thumb mode */
 		tracee->_regs[CURRENT].ARM_cpsr &= ~PSR_T_BIT;
+#endif
 #endif
 
 		/* Restore registers to their current values.  */
